[Draft PR]Rewrite Thingiverse Provider API Script to use new ImageStore class

aashrafh opened the pull request:
<!-- Please replace #XX below with an existing issue number. Remove the line entirely if none exist. -->
Fixes #252

## Description
In this draft, I nearly finished the code but it needs to be reviewed and write e test for it. 
What I have finished:
* The new script uses the new ```ImageStore``` class.
* The new script uses the new ```DelayedRequester``` class.
* The new script does not use anything from ```src/cc_catalog_airflow/dags/provider_api_scripts/modules/etlMods.py```.
* The new script provides a main function that takes a date parameter of the form ```YYYY-MM-DD```.
* The new script conforms to PEP8.

## Other information
Currently,  I am working on the test script for the new script along with reviewing the new script for more optimizations and any bugs.

## Checklist
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the `master` branch of the repository.
- [x] My commit messages follow [best practices][best_practices].
- [x] I added tests for the changes I made (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

<!-- Make sure you read and understand the following attestation. -->
## Developer Certificate of Origin
```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

aashrafh committed:
Rewrite Thingiverse Provider API Script to use new ImageStore class

mathemancer reviewed (changes requested):
For the most part, this is great!  Please see my specific comments and the checkboxes for changes.  

In order to get the script to run on your local machine, you'll need to get your own API token from Thingiverse (see https://www.thingiverse.com/developers/getting-started).  If that causes trouble for you, let me know and we'll figure out how to proceed.

Luckily, the way you've separated out logic means you'll be able to test many things without calling the actual API.  Please begin writing some unit tests, and let me know where you're blocked by API Access if you can't get that working.

- [ ] Remove unused imports, variables, and constants
- [ ] Run `pycodestyle` on your code and fix anything it finds.  The exception to this is long lines in test files due to long strings.
- [ ] Make sure the script will run on your machine (this one can't load in its current form)

Thank you!

aashrafh commented:
@mathemancer thanks for fast review. I will start working on the changes considering the verifying of running the script locally. I will back with you very soon

aashrafh committed:
Remove unused imports, variables, and constants

aashrafh committed:
Python style guide checking using pycodestyle

aashrafh committed:
fix a typo

aashrafh committed:
move license and license_version to constants

aashrafh committed:
hit the tags endpoint to get the tags list

aashrafh22 committed:
extract bath thing process logic

aashrafh commented on src/cc_catalog_airflow/dags/provider_api_scripts/thingiverse.py:
```diff
@@ -0,0 +1,372 @@
+                    thing, start_timestamp, end_timestamp)
+                if total_images == 0:
+                    is_valid = False
+                    break
```
please review the last commit I made

aashrafh22 committed:
fix creator bug

aashrafh commented on src/cc_catalog_airflow/dags/provider_api_scripts/thingiverse.py:
```diff
@@ -0,0 +1,372 @@
+                thing_data['creator']['last_name'])
+
+        if (creator.strip() == '') and ('name' in thing_data['creator']):
+            creator = thing_data['creator']['name']
```
I'm not sure if understood you correctly but I made a commit and hope it will fix the bug. Please, review it.

aashrafh replied:
Also, try to run the script now. It's supposed to run successfully after the last commit

aashrafh22 committed:
fix the bugs that prevent the script from running

aashrafh22 committed:
initial test code

aashrafh22 committed:
Complete test_thingiverse and fixes some bugs in thingivers.py

mathemancer reviewed (changes requested):
First, and foremost, you have committed and pushed your Thingiverse token to Github.  I strongly suggest you invalidate your Thingiverse Token, get a new one, and avoid committing it or pushing it to Github in the future.  Don't feel too bad about this, I've made a similar mistake in the past.

Also, the script is not running on my machine.  The Logs are:
```
2020-03-03 13:56:39,843 - common.storage.image - INFO:  Initialized with provider thingiverse
2020-03-03 13:56:39,843 - common.storage.image - INFO:  No given output directory.  Using OUTPUT_DIR from environment.
2020-03-03 13:56:39,843 - common.storage.image - INFO:  Output path: /tmp/thingiverse_20200303135639.tsv
2020-03-03 13:56:39,843 - __main__ - INFO:  Processing Thingiverse API for date: 2020-03-01
2020-03-03 13:56:39,844 - common.requester - INFO:  Processing request for url: https://api.thingiverse.com/newest
2020-03-03 13:56:39,844 - common.requester - INFO:  Using query parameters {'access_token': '32044862bce84c5399505e5c85d40c2a', 'per_page': 30, 'page': 1}
2020-03-03 13:56:39,844 - common.requester - INFO:  Using headers None
Traceback (most recent call last):
  File "thingiverse.py", line 392, in <module>
    main(date)
  File "thingiverse.py", line 60, in main
    cur_page
  File "thingiverse.py", line 98, in _get_things_batch
    response_json = _get_response_json(query_params, retries=retries)
  File "thingiverse.py", line 148, in _get_response_json
    response_json.get('error') is not None
AttributeError: 'list' object has no attribute 'get'
```
Finally, there are some typos in naming of directories and test resource files preventing the tests from running on my machine.

Please fix these issues, and re-request review from me.

P.S., please don't be discouraged, and sorry it took me so long to get to this PR review.

mathemancer commented:
See https://stackoverflow.com/questions/1270514/undoing-a-git-push to remove the Token from your git history.

aashrafh22 committed:
thingiverse.py and test_thingiverse.py

aashrafh commented:
@mathemancer, sorry about the mistake I made.
I have rewritten some of ```thingiverse.py``` and tested all the functions using ```test_thingiverse.py``` but I have a problem.
The problem is everything is working correctly in the unit testing but when I run ```thingiverse.py``` it runs successfully but no output! I have tested many things and rewritten many things but still, same problem so any help or suggestions?

aashrafh22 committed:
fix PATH in etlMods.py

mathemancer commented:
>  The problem is everything is working correctly in the unit testing but when I run `thingiverse.py` it runs successfully but no output! I have tested many things and rewritten many things but still, same problem so any help or suggestions?

It appears `_process_image_list` is deleting everything, and so every time `_add_images` is called, it's being called with an empty list.  Let me know if this helps!

mathemancer reviewed (changes requested):
Please see my earlier comment.  Please re-request review once the bug resulting in always calling `_add_images` with an empty list is solved.

Thank you!

aashrafh22 committed:
fix process image bug

aashrafh commented:
Hello @mathemancer, I found that the problem is not in the ```_process_image_list``` itself but the problem is ```cur_page```. I wrote a unit test for the ```_process_image_list``` function and found that it works correctly so I tried somethings manually with the API. I found that nearly first 3000 pages of this endpoint ```https://api.thingiverse.com/newest``` the list images JSON is different from the other old lists. To test what I say try to hit the ```files``` endpoint of the thing```3949469``` and the thing ```3304806``` you will see the difference.
So, what I did so far: 
* Changes ```cur_page``` value to suitable value to hit the right endpoint but this is a temporary solution till I write some code to work properly on both the old and new lists.
* I added new constant ```IMAGES_LIMIT``` to limit the number of images as there so many images can be pulled which will consume much time. 
Finally, the code in the current state is not the best that can be achieved but it little messy as I was trying to figure out the problem so I will consider organizing it.

aashrafh commented:
Also, there are two ```.tsv``` files in ```test/resources/thingiverse```. Both of them generated successfully by ```thingiverse.py```

mathemancer commented:
Very interesting...So the newer things don't come with a `default_image` field, or am I missing the point?

aashrafh22 committed:
Example of null response

aashrafh commented:
I tested the problem again today to see the recently added things and noticed that the problem actually is ```default_image``` field is ```null``` for many recent pages on Thingiverse. I tried on different pages like pages 33, 52, 1000, 2000, and 3000 and the same result all of them are ```null``` value for the ```default_image```.

I added three examples of this null response so you can review it. I can conclude the problem is the empty list form ```_process_image_list``` was actually expected.

aashrafh committed:
Merge branch 'master' into master

aashrafh22 committed:
get rid of unnecessary code and run the script to test it

aashrafh22 committed:
Merge branch 'master' of https://github.com/aashrafh/cccatalog

mathemancer commented:
> I tested the problem again today to see the recently added things and noticed that the problem actually is `default_image` field is `null` for many recent pages on Thingiverse. I tried on different pages like pages 33, 52, 1000, 2000, and 3000 and the same result all of them are `null` value for the `default_image`.

This does seem to be the case.  I think I'll have to move this ticket to 'blocked' status pending figuring out why that's the case.  There's not much point in spending effort pulling images from Thingiverse if there aren't any.

aashrafh commented:
That's fine. Once this problem is solved I'm available to finish this issue.

mathemancer commented:
Unfortunately, I'll have to close this due to the pausing of the project.
